### ISSUE TITLE
Ideas for memory management for large initrd ramdisks

### ISSUE THREAD

**Issue by ghaerr** [mikaku/Fiwix#80]
Hello @mikaku and @rick-masters,

I'm opening a new issue to continue discussion from https://github.com/mikaku/Fiwix/issues/78#issuecomment-1963025600 so that the topic is more easily followed. I'm repeating below that last post for completeness here:

> > Can you explain in a bit more detail what the problem is here? It seems to me that MultiBoot first loads the optional initrd into physical RAM just after the kernel text/data (just before kpage_dir). Then mem_init remaps that initrd RAM above PAGE_OFFSET before loading the final kernel GDT, right?
> > So how is initrd taking up user space virtual addresses? Aren't all the initrd addresses in KVA?

> Initially the Fiwix kernel only had the virtual memory split 3/1 (3GB user / 1GB kernel) and so the maximum available physical memory was 1GB, regardless if your PC had more memory. In this case the RAMdisk drives or the initrd were limited to a maximum of around 950MB, but if you tried to use all the memory, you cannot even login to the system because there would not be enough memory for the user applications.
>
> In https://github.com/mikaku/Fiwix/issues/34, @rick-masters suggested a patch to increase the limit of the size in the initrd images to be more than 1GB, which is a requirement for their Bootstrappable project. But somehow we agreed that the patch was a bit tricky.
>
> So, in my to attempt to fix this, I presented a new path, the support to have the virtual memory split 2/2 (2GB user / 2GB kernel) with the new kernel option CONFIG_VM_SPLIT22 (disabled by default). This way you can have easily an initrd of 1.2GB in size and you still have around 750MB for the user space.
>
> Recently though, people in the Bootstrappable project complained that with very large initrd files (let's say 1.5GB) they only have around 450MB for user, even when the system has 4GB of physical memory. They ask if there would be a change to use the unused 2GB of physical memory for the initrd files and the rest for the user space.

Thanks for the background information, very interesting.

> I continue thinking how can I implement this in a decent way (no tricky) but so far I'm failing on this.

If I understand the problem correctly, Fiwix can be configured for 2/2 (2GB user / 2GB kernel) or 3/1 (3GB user / 1 GB kernel) but the kernel virtual address space size is the limiting factor for the amount of physical memory managed, as currently all physical memory is permanently mapped into kernel VA, including initrd/ramdisks.

Thus, with a 2/2 system and 1.5G initrd, the max physical memory allowed and mappable is 2GB, which then leaves only .5GB left for user space (less kernel text/data). Likewise, with a 3/1 system, max physical RAM is 1GB with even less ramdisk and user space available.

Why do RAMdisks have to be permanently mapped into KVA? They're a normal block driver, and accessed only through the FS ops table. It would seem that the ramdisk driver could be easily updated to use a pre-specified set of PTEs, say a single PT managing 4MB of memory, that could be updated by the ramdisk driver routine to access ramdisk data outside of otherwise mapped kernel memory. This memory would be allocated directly above _last_data_addr, and not be included in kstat.physical_pages.

Depending on the total actual physical RAM available, the kernel could allocate what is needed or specified for user space in kstat.physical_pages, then use addresses just above that for RAMdisks, using a single PT to access the entire ramdisk addressable range for caching purposes. Of course, any Multiboot initrd would have to be copied from lower memory once, but this could be done by the ramdisk init routine, leaving the lower memory available for user space and not marked PAGE_RESERVED.

The ramdisk PTEs would operate very much like how the framebuffer mmap now works, by containing an otherwise unknown-to-kernel physical memory address (along with the PAGE_NOALLOC bit, although not strictly needed).

This design removes all ramdisk contents from the physical memory previously managed by Fiwix, at the cost of possibly updating PTE entries outside the "cached" (or should I call it "bank-switched") window into ramdisk contents. Since most frequently accessed ramdisk contents would be cached in the buffer system anyways, the benefit of having much more memory available for user space outweighs any speed decrease, it would seem.

Does this idea sound worth pursuing? What do you think?

**Comment by mikaku** [comment]
> If I understand the problem correctly

Yes, you understand it correctly.

> Why do RAMdisks have to be permanently mapped into KVA? 

Yes, well, my explanation was not complete, sorry. The directory `docs/` contains some files with details about memory layout. Indeed RAMdisk drives are fixed memory allocated during kernel boot up, but that's not the problem. The people in Bootstrappable project needs a biginitrd file, I mean a file that can be bigger than 1.3GB of size. The initrd files are like RAMdisk drives but they are previously loaded by the bootloader in a fixed memory position.

With initrd files you cannot use the dynamic allocation like you could do on RAMdisk drives. Effectively, I already have planned to changed the way how the RAMdisk drives are allocated in the kernel. I plan to use the buffer cache on top of the RAMdisk drives, like Linux does. So Fiwix don't need to preallocated the whole memory of a RAMdisk, it will be allocated on demand as the user fills it up.

So, what I tried to explain is that RAMdisk drives aren't really a problem here. The problem is the initrd file (which is also a RAMdisk drive, but not a generic one), that can has a size of 1.3GB or more and must be memory preallocated.

> This memory would be allocated directly above `_last_data_addr`, and not be included in kstat.physical_pages.

Yes, I already thought on this, but I'm not sure if this would work.

> This design removes all ramdisk contents from the physical memory previously managed by Fiwix, at the cost of possibly updating PTE entries outside the "cached" (or should I call it "bank-switched") window into ramdisk contents. Since most frequently accessed ramdisk contents would be cached in the buffer system anyways, the benefit of having much more memory available for user space outweighs any speed decrease, it would seem.

I'm not sure if I understand it correctly, excuse my ignorance. It looks like you have a nice idea of how implement this and I wouldn't be a stopper here.

I know that you are currently focused in your Microwindows port and the rest, but I'd like to resume this idea once you have completed the Microwindows tasks, and your mind will be free to focus on this. :-)

**Comment by ghaerr** [comment]
> The people in Bootstrappable project needs a biginitrd file, I mean a file that can be bigger than 1.3GB of size. The initrd files are like RAMdisk drives but they are previously loaded by the bootloader in a fixed memory position.

Yes - the idea would be to copy initrd's physical memory previously allocated by the boot loader to a new location above the memory occupied by kstat.physical_pages, thus releasing the PAGE_RESERVED memory initrd now uses to be available for user space mapping, and using a page table managed by the initrd block driver to access the initrd data in high memory, otherwise outside of use by the kernel.

> So, what I tried to explain is that RAMdisk drives aren't really a problem here. The problem is the initrd file (which is also a RAMdisk drive, but not a generic one), that can has a size of 1.3GB or more and must be memory preallocated.

Yes - but an initrd uses drivers/block/ramdisk.c as its driver, right? So I am only really talking about initrd here, but it could also be applied to ramdisks in general. Since the preallocated size is known, it can be copied outside of the normal memory used by Fiwix very easily to higher memory physical pages that won't be used for user space (nor marked PAGE_RESERVED). 

> I plan to use the buffer cache on top of the RAMdisk drives, like Linux does. So Fiwix don't need to preallocated the whole memory of a RAMdisk, it will be allocated on demand as the user fills it up.

If initrd uses the same block driver as ram disks, then using the buffer cache will also require kernel page-mapped buffers (that is, kernel-managed memory) for the disk contents. This will then reduce the availability of the page cache for use in user space virtual addresses, which I also thought was part of the problem. (That is, large initrd takes away from user virtual address space).

> I'm not sure if I understand it correctly, excuse my ignorance. It looks like you have a nice idea of how implement this and I wouldn't be a stopper here.

Basically all I was saying is to not use any memory below kstat.physical_pages for initrd at all; instead to copy the boot-loaded initrd to a high physical address above kstat.physical pages and use a single page table in the initrd block driver to access it, outside of all other normal page management. This single page table would be updated as required to access initrd contents via the read/write block entry point in the driver. 

The initrd/(ramdisk) driver would essentially hide/encapsulate its dedicated page table updates outside of all other kernel page handling, as its memory would not be part of the kernel-managed memory currently used.

> but I'd like to resume this idea once you have completed the Microwindows tasks, and your mind will be free to focus on this. :-)

Yeah, no problem. I wasn't thinking of jumping in to coding but just talking about it, since discussions like this help me understand Fiwix memory and page management. The Microwindows port is done, and only requires a single config change to build client/server, which I will test shortly using Newlib 4.2.

Thank you!

**Comment by mikaku** [comment]
> Basically all I was saying is to not use any memory below kstat.physical_pages for initrd at all; instead to copy the boot-loaded initrd to a high physical address above kstat.physical pages and use a single page table in the initrd block driver to access it, outside of all other normal page management. This single page table would be updated as required to access initrd contents via the read/write block entry point in the driver.

The initrd/(ramdisk) driver would essentially hide/encapsulate its dedicated page table updates outside of all other kernel page handling, as its memory would not be part of the kernel-managed memory currently used.

I believe you have a much clear idea than me. I think that the initial patch of @rick-masters already reallocated the initrd to a higher memory address, but I'm not sure.

Also, the idea must be consistent in the case when the 'above' memory is not enough for the initrd, in that case, should the kernel refuse to map the initrd file (or RAMdisk drive)? I mean, reallocating the initrd to upper addresses is a great idea, but it will be consistent in all cases?

Well, that's something that we can discuss later.

**Comment by ghaerr** [comment]
> Also, the idea must be consistent in the case when the 'above' memory is not enough for the initrd, in that case, should the kernel refuse to map the initrd file (or RAMdisk drive)? I mean, reallocating the initrd to upper addresses is a great idea, but it will be consistent in all cases?

Agreed it should work when there is not enough "above" memory for initrd. In that case, instead of refusing to run, kstat.physical_pages would be adjusted downwards by the amount required, which would still allow the initrd to be mapped highest, with the net result of reducing the available memory for user space virtual addresses, just like it does now. Here is where `bios_map_init` initializes  kstat.physical_pages and possibly truncates it:
```

    /*
     * This truncates to 1GB since it's the maximum physical memory
     * currently supported.
     */
    if(kstat.physical_pages > (GDT_BASE >> PAGE_SHIFT)) {
        kstat.physical_pages = (GDT_BASE >> PAGE_SHIFT);
        printk("WARNING: only up to %dGB of physical memory will be used.\n", GDT_BASE >> 30);
    }
```
Instead of truncating there, `mem_init` would adjust it downwards by the amount required for the initrd to be placed at the top of physical RAM.

In effect, the design moves initrd out of the middle of physical RAM to the top of it, which, if there's enough physical RAM, allows the Fiwix kernel to map a larger user address space within the paged memory AND map initrd memory outside of the page-managed area above _last_data_area. By not keeping the initrd RAM permanently mapped in the page tables, a larger total physical address space can be managed.

**Comment by mikaku** [comment]
If I understand correctly your idea, the kernel will reserve a page table (4KB) to map 4MB of any position of the `/dev/ram0`, (and another page table for `/dev/ram1`, and so on). Such page table will be used on the standard semantics (open, read, write, close), and also during `mmap()` calls.

If so, I think it may work, but it will require some kind of protection if multiple users want to read/write on the same RAMdisk drive. And also, more protection will have to exist in the future when implementing SMP support.

Does this make sense to you?

**Comment by mikaku** [comment]
> then a full 1024 PTEs would have to be rewritten, but not used. This is possibly too much overhead;

Hmm, I think that the best solution is to maintain the current RAMdisk memory implementation and create a new kernel option that would activate your idea if the user needs to use the third and forth GB of memory, at the cost of having a bit more overhead.

This way, both mechanisms would coexist in the code. In fact, the majority of people will use the current RAMdisk memory implementation which is good for almost all cases.

**Comment by ghaerr** [comment]
> I think that the best solution is to maintain the current RAMdisk memory implementation and create a new kernel option that would activate your idea

Well, that could essentially be a second RAMdisk driver implementation, which would replace the first.

> In fact, the majority of people will use the current RAMdisk memory implementation which is good for almost all cases.

Given that there's more to do than just replace the RAMdisk driver, especially the issues regarding figuring out which portions of the third and fourth GB of memory would be used, the BIOS mapping changes, initrd copy, and the possible `#ifdef`s that might be brought in, I suppose there's no point in implementing it if nobody really cares.

I had thought that the issue might have been more important and I like solving user problems. Given that Fiwix is mostly designed and written to be very easy to learn and understand, I suppose a feature like this should probably wait until there are other needs in Fiwix to take advantage of more physical memory. It was fun coming up with the idea of introducing non-permanently mapped kernel virtual address space nonetheless. Thank you!

**Comment by mikaku** [comment]
Don't get me wrong. I think your idea is really good and I believe that it could work. On the other hard, I don't want you to do an incredible effort for something that will benefit just a few users. We can resume this idea later, if there is a solid demand for it.

Sincerely, I'm more excited to include support for a TCP/IP stack and Nano-X (or even X11) and with all the packages that will be need to port.

**Comment by ghaerr** [comment]
>  I'm more excited to include support for a TCP/IP stack and Nano-X (or even X11)

My last work on Nano-X showed that UNIX sockets have some problems when multiple clients are reading/writing to a server waiting on `select`. It would be nice to have a non-graphical test program to test and identify any problem(s). Do you have your UNIX sockets test suite code published somewhere? That could be used as a start and possibly extended to help with this problem. We'll definitely need sockets working well before tackling TCP/IP or an X11 server.

**Comment by mikaku** [comment]
>  Do you have your UNIX sockets test suite code published somewhere? 

Yes, but all the example tests I have collected from Internet are using simple server-client connections without using `select()` and they are all working fine. The only decent example using `select()` I found is [here](https://www.smnd.sk/anino/programming/c/unix_examples/poll.html) (discovered recently). This is a perfecte example to see that the `select()` implementation doesn't work in UNIX sockets under Fiwix.

Both programs (client and server) compile successfully, but you better redefine the name of the `SOCKETNAME` constant.

I recommend you to open the `ttyS0` and `ttyS1` devices by modifying the `/etc/inittab` file, so you can run `./pollsrc` in console and the two `./pollcli` from serial lines.

You might want to test first these programs under Linux to familiarize how they work.

**Comment by ghaerr** [comment]
> The only decent example using select() I found is [here](https://www.smnd.sk/anino/programming/c/unix_examples/poll.html)

After fixing the following bug and adding a couple standard header files, I was able to get the two programs running on macOS for testing, which then worked:
```
    len = sizeof(name.sun_family) + strlen(name.sun_path) + 1; // <--- Add 1 for termination NUL required
```
The same programs were successfully compiled on Fiwix, and I found that the client/server does work, except that only the LAST accepted connection (that is, the last connected ./pollcli) was able to send messages. So "one" UNIX socket connection works, but apparently only the second one... Rather than look everywhere, I would like to ask you why you think that might be?

> I recommend you to open the ttyS0 and ttyS1 devices by modifying the /etc/inittab file, so you can run ./pollsrc in console and the two ./pollcli from serial lines.

What is the qemu command line to get that to work? I was able to test by running `./pollsrv &` and then running two clients, one on the console, and the other on /dev/ttyS0 using the `qemu -serial stdio` option. I don't know who one might configure two serial lines using QEMU?

**Comment by mikaku** [comment]
> I would like to ask you why you think that might be?

Not sure yet. I need to do more tests.

> What is the qemu command line to get that to work? 

I include the following lines to have `/dev/ttyS0` and `/dev/ttyS1` on FiwixOS:

```
qemu-system-i386 \
        ...
        -chardev pty,id=pciserial \
        -device pci-serial,chardev=pciserial \
        -serial pty
```

Then it's just a matter to open two terminals on your macOS and run `picocom --imap lfcrlf /dev/pts/[n]` on each one. You'll know `/dev/pts/[n]` because QEMU will show you a pair of similar lines when you run it from a terminal:

```
$ qemu-system-i386 ...
char device redirected to /dev/pts/7 (label pciserial)
char device redirected to /dev/pts/8 (label serial0)
```

I hope that helped you.

**Comment by mikaku** [comment]
By the way, I've reorganized the repository FiwixOS and includes the instructions to be able to build your own packages.
Please, let me know if the README.md file is clear enough.

**Comment by ghaerr** [comment]
> let me know if the README.md file is clear enough.

Yes, much improved, thanks. I am still a bit confused as to when to run `makeall.sh` vs `make_media.sh`.

Does `makeall.sh` rebuild everything from scratch, or does it just update the changed packages whose source has changed? What happens if a new header file is added to the C library, does that require a full rebuild?

I am guessing that `make_media.sh` is used to create an ISO file (somewhere/somehow?), pointed to the source files? The example shows `/path/to/fiwix/source/code` but not sure what that actually is supposed to be changed to. 

And `install.sh` is only required for actually installing FiwixOS onto a separately mounted filesystem, instead of when booting from `FiwixOS-3.3-i386.raw`, right?

Of course, none of the host instructions will work on macOS, as it knows nothing about EXT2 filesystems nor has any utilities for such. If you like, I can provide instructions for how to transfer the FiwixOS repo to a filesystem Fiwix can use, except that for the moment, I only have working utilities for MINIX2 filesystem (using the `mfs` tool from ELKS).

**Comment by ghaerr** [comment]
> I found that the client/server does work, except that only the LAST accepted connection (that is, the last connected ./pollcli) was able to send messages. 

More testing shows my comment above is not quite correct. What happens is that both connection are in fact accepted, but very strangely, while the last connection reads from stdin and through the server every time, the first connection "delays" the first two (yes, exactly two) reads, and they are then delivered, in order, without data loss, with each subsequent read (while those reads are queued for later delivery). That is, the first connection seems to have its data delayed for delivery to the other client by two messages!!

The reason I was initially confused is that I just hit <return> twice on the first client, which does nothing. However, on the next (third) read of stdin, the first-read data is in fact delivered, ad infinim with each subsequent stdin read.

What is most strange about this is that, looking at the UNIX sockets code, one would think that the messages would get combined if they were not delayed (that is, delivered all together as a single message). Thus I would guess that perhaps the problem is not in the UNIX sockets code directly, but in the scheduling mechanism somehow in between the console reads and delivery to of from UNIX sockets? Not sure, but wanted to report this finding.

**Comment by ghaerr** [comment]
Figured it out!!! Fix coming for UNIX sockets support.

Nano-X is now working with multiple clients. Yay!!! :)

**Comment by mikaku** [comment]
Sorry for the late reply, I've been on a trip since early Friday.

(wow!, just read your last post. Good job!)

> I am still a bit confused as to when to run makeall.sh vs make_media.sh.

Yes, with `makeall.sh` you can build all the packages currently ported. This script stops after each build, but you can include the parameter `y` to do a continue build.

The script `make_media.sh` is not functional right now due to the reorganization I made in the directories. It servers to create the FiwixOS media (floppy disk, Installation CDROM, etc.).

> What happens if a new header file is added to the C library, does that require a full rebuild?

The GNU Toolchain and the Newlib C are built on a separate directory with its own script `make-toolchain.sh`. Every time you want to include a new header file to Newlib C, you can include it also in your own FiwixOS root filesystem, and postpone the rebuild of the Newlib C to do it for a more serious change (a new syscall, etc.).

> And install.sh is only required for actually installing FiwixOS onto a separately mounted filesystem, instead of when booting from FiwixOS-3.3-i386.raw, right?

`The script `install.sh` is actually the FiwixOS installation script. It is copied into the Installation CDROM when the media is created (using the script `make_media.sh`. You don't need to touch the `install.sh` script unless you really want to change how FiwixOS is installed. I don't plan to make big changes on this script in the near future.

>  If you like, I can provide instructions for how to transfer the FiwixOS repo to a filesystem Fiwix can use

Feel free to include a README file to include the steps needed to do this.

> What is most strange about this is that, looking at the UNIX sockets code, one would think that the messages would get combined if they were not delayed (that is, delivered all together as a single message). Thus I would guess that perhaps the problem is not in the UNIX sockets code directly, but in the scheduling mechanism somehow in between the console reads and delivery to of from UNIX sockets? Not sure, but wanted to report this finding.

I made a first test with the example in the URL above and I saw immediately that the communication with both clients didn't work at all. I wanted to have some time to try to find the cause of this malfunction, but it seems you did it again! :-)

> Figured it out!!! Fix coming for UNIX sockets support.
Nano-X is now working with multiple clients. Yay!!! :)

This is amazing, I thought this bug would be a big nightmare. Great job Gregory!!
:-)

**Comment by ghaerr** [comment]
> I think that the best solution is to maintain the current RAMdisk memory implementation and create a new kernel option that would activate your idea if the user needs to use the third and forth GB of memory, at the cost of having a bit more overhead.

> In fact, the majority of people will use the current RAMdisk memory implementation which is good for almost all cases.

After more consideration on the issue of copying a large Multiboot-loaded initrd from the middle of memory to higher memory, I realized there might not be enough RAM to do such a thing. That case would then require a "fallback" to the current implementation. In addition, I have learned the current KEXEC code for Linux is somewhat intertwined with having direct memory access to (at least the initial) RAMdisk contents.

Thus, until the idea that Fiwix might support "memory zones" similar to Linux's ZONE_NORMAL and ZONE_HIMEM along with its ensuing complications, these memory management and initrd/RAMdisk driver enhancements probably go against Fiwix's primary goal of being educational and easy-to-understand, rather than being built to maximize physical use of memory on real hardware, so I'll close this issue now.

**Comment by mikaku** [comment]
I've found [this](https://lkml.org/lkml/2004/10/9/126) old Linux kernel Mailing List email that can be of help if we ever retake this idea:

**Comment by ghaerr** [comment]
> this old Linux kernel Mailing List email that can be of help if we ever retake this idea

Yes, my take on that kind of patch for Fiwix would be that 1) for the simplest case of using additional CONFIG_SPLIT_xxx options, it could be interesting for a much larger kernel data address space at the expense of user space if MMAP_START (0x40000000) was also adjusted downwards for each of the lower user space options.

Or, a more complex option 2) would be the implementation of `vmalloc` for temporarily mapping Linux-style "high memory" pages into upper kernel VA with Linux 2.0 style VMALLOC_START and VMALLOC_OFFSET and then allowing those pages to be mapped into a user process, possibly with more complex page fault handling when memory is tight.

As you said before, it really depends on what users actually require; I didn't really get a good understanding of what the #liveboot project required other than a very large initrd/ramdisk (that is, do they also need a large user space in addition?) 

Should Fiwix need it, I have studied enough to bring a `vmalloc` to life along with high_memory page usage like Linux, but am still struggling with exactly when and for how long those pages would be allocated to either kernel or user space.

**Comment by mikaku** [comment]
> As you said before, it really depends on what users actually require; I didn't really get a good understanding of what the #liveboot project required other than a very large initrd/ramdisk (that is, do they also need a large user space in addition?)

Yes, the guys in the project [Live Bootstrap](https://github.com/fosslinux/live-bootstrap/blob/master/parts.rst) (with its own Libera.Chat channel #bootstrappable need to create a really big initrd image (around 1.3GB) to accomplish a complete build. But don't ask me why because I'm not directly part of this project. During almost all 2023 year I collaborated with them accepting plenty of PRs to improve the Fiwix kernel in order they can use the official version, instead of running their own fork.

> Should Fiwix need it, I have studied enough to bring a vmalloc to life along with high_memory page usage like Linux, but am still struggling with exactly when and for how long those pages would be allocated to either kernel or user space

Considering that besides you and me, I don't know of other users using Fiwix, we can postpone this memory implementation, and focus on different aspects.

**Comment by mikaku** [comment]
> Yes, my take on that kind of patch for Fiwix would be that 1) 

Agreed, PAGE_OFFSET and MMAP_START would have to be adjusted for each split implemented.

> Or, a more complex option 2)

This would be the right path to follow when including support for the amd64 architecture, since the emory implementation will have to change a lot.
